newrelic/hooks/mlmodel_openai.py

```diff
@@ -0,0 +1,145 @@
+        "response.model": response_model,
+        "response.organization":  response.organization,
+        "response.usage.completion_tokens": response.usage.completion_tokens,
+        "response.usage.total_tokens": response.usage.total_tokens,
+        "response.usage.prompt_tokens": response.usage.prompt_tokens,
+        "request.temperature": kwargs.get("temperature", ""),
+        "request.max_tokens": kwargs.get("max_tokens", ""),
+        "response.choices.finish_reason": response.choices[0].finish_reason,
+        "response.api_type": response.api_type,
+        "response.headers.llmVersion": response_headers.get("openai-version", ""),
+        "response.headers.ratelimitLimitRequests": check_rate_limit_header(response_headers, "x-ratelimit-limit-requests", True),
+        "response.headers.ratelimitLimitTokens": check_rate_limit_header(response_headers, "x-ratelimit-limit-tokens", True),
+        "response.headers.ratelimitResetTokens": check_rate_limit_header(response_headers, "x-ratelimit-reset-tokens", False),
+        "response.headers.ratelimitResetRequests": check_rate_limit_header(response_headers, "x-ratelimit-reset-requests", False),
+        "response.headers.ratelimitRemainingTokens": check_rate_limit_header(response_headers, "x-ratelimit-remaining-tokens", True),
+        "response.headers.ratelimitRemainingRequests": check_rate_limit_header(response_headers, "x-ratelimit-remaining-requests", True),
+        "vendor": "openAI",
+        "ingest_source": "Python",
+        "number_of_messages": len(kwargs.get("messages", [])) + len(response.choices),
+    }
+
+    transaction.record_ml_event("LlmChatCompletionSummary", chat_completion_summary_dict)
+    message_list = list(kwargs.get("messages", [])) + [response.choices[0].message]
+
+    create_chat_completion_message_event(transaction, settings.app_name, message_list, chat_completion_id, span_id, trace_id, response_model, response_id, request_id)
+
+    return response
+
+
+def check_rate_limit_header(response_headers, header_name, is_int):
+    if not response_headers:
+        return ""
+
+    if header_name in response_headers:
+        header_value = response_headers.get(header_name)
+        if is_int:
+            header_value = int(header_value)
+        return header_value
+    else:
+        return ""
+
+
+def create_chat_completion_message_event(transaction, app_name, message_list, chat_completion_id, span_id, trace_id, response_model, response_id, request_id):
+    if not transaction:
+        return
+
+    for index, message in enumerate(message_list):
+        chat_completion_message_dict = {
+            "id": "%s-%s" % (response_id, index),
+            "appName": app_name,
+            "request_id": request_id,
+            "span_id": span_id,
+            "trace_id": trace_id,
+            "transaction_id": transaction._transaction_id,
+            "content": message.get("content", "")[:MAX_LOG_MESSAGE_LENGTH],
+            "role": message.get("role"),
```

Comment:
Should probably set a default for this.
```suggestion
            "role": message.get("role", ""),
```